d2compiler/compile.go

```diff
@@ -735,8 +735,18 @@ func (c *compiler) validateNear(g *d2graph.Graph) {
 					c.errorf(obj.Attributes.NearKey, "constant near keys can only be set on root level shapes")
 					continue
 				}
-				if len(obj.ChildrenArray) > 0 {
-					c.errorf(obj.Attributes.NearKey, "constant near keys cannot be set on shapes with children")
+
+				descendantsMap := getNearDescendants(obj)
+				connectToOutside := false
+				for _, edge := range g.Edges {
```

Comment:
simplify. loop through edges and make sure that the near ancestor of both endpoints match. no need for the extra map. 

i know it's less efficient, but it's negligible and simplifies code